### PR TITLE
In-memory engine: clear write batch when range load failed

### LINKED ISSUE #17103: In-memory Engine: cached_write_batch is not cleared after load failure
## Bug Report

cached_write_batch is not cleared when load failed so that when the next time the same range is to load, dirty data sill be written.

For lock cf, it can cause lock disappear or reappear which results in data inconsistency.

### What version of TiKV are you using?

build on [215484e](https://github.com/tikv/tikv/pull/16868/commits/215484e84437b132f7d0db8a20ee907e65d1678f)

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #17103 Ref #16141

What's Changed:

```commit-message
clear write batch when range load failed
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/region_cache_memory_engine/src/background.rs [hunk 1]
```diff
@@ -439,6 +439,7 @@ impl BackgroundRunnerCore {
     // if `false` is returned, the load is canceled
     fn on_snapshot_load_finished(&mut self, range: CacheRange) -> bool {
         fail::fail_point!("on_snapshot_load_finished");
+        fail::fail_point!("on_snapshot_load_finished2");
```
Comment (zhangjinpeng87) [inline components/region_cache_memory_engine/src/background.rs:442]: What is the difference between these above two fail points?

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/background.rs:442]: In the test in this PR, one is used to notify the execution is reached (by cfg_callback) and the other is used to pause the execution.

Comment (LykxSassinator) [reply components/region_cache_memory_engine/src/background.rs:442]: How about using the bool flag as the trigger just like:
```
    fail_point!(
            "on_snapshot_load_finished_range",
            range != CacheRange(start_key, end_key),
            |_| {}
        );
```

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/background.rs:442]: The notification and pause I mentioned above is for the same range.

**Review by overvenus — APPROVED** [review]

**Review by LykxSassinator — APPROVED** [review]

**Pull request merged**